### PR TITLE
Issue#166 fall-back for incomplete CWE's (race)

### PR DESCRIPTION
**Pull request by cragun47**
This provides a fall-back for incomplete or unknown CWE's according to the outlined solution.  Unit tests added.

Note, a formatter was used on the files, so you may see more lines flagged as changed than the actual code changes.

### CODE REVIEW

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 1]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
-
     }
 
     @Test
     void test_that_extension_handles_alternate_CDCREC_race_encoding() {
 
         String patientWithDataForExtensions = "MSH|^~\\&|MIICEHRApplication|MIIC|MIIC|MIIC|201705130822||VXU^V04^VXU_V04|test1100|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS|^^^^^MIIC^SR^^^MIIC|MIIC\n"
                 // Test text only race
-                + "PID|1||12345678^^^^MR|ALTID|Mouse^Mickey^J^III^^^|||M|Alias^Alias|2106-3^White^CDCREC|USAA|||english|married||||DL00003333|||Orlando Disney Hospital|Y|2|USA||||\n";
+                + "PID|1||12345678^^^^MR|ALTID|Mouse^Mickey^J^III^^^|||M|Alias^Alias|2106-3^White^CDCREC||USAA|||english|married||||DL00003333|||Orlando Disney Hospital|Y|2|USA||||\n";
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:123]: We've been asking that HL7 messages be simplified to make the tests targetted, and clearly de-identified.  For this one I suggest:
`PID|1||12345678^^^^MR||Doe^Jane|||||2106-3^White^CDCREC`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 2]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        CodeableConcept ccW = (CodeableConcept) extensions.get(0).getValue();
+        assertThat(ccW.hasCoding()).isTrue();
+        Coding coding = ccW.getCodingFirstRep();
+        assertThat(coding.hasDisplay()).isTrue();
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getDisplay()).hasToString("White");
+        assertThat(coding.getCode()).hasToString("2106-3");
+        assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
+
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:141]: Add checks:
```
assertThat(ccW.hasText().isTrue();
assertThat(ccW.getText().hasToString("White");
```

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 3]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+    }
+
+    @Test
+    void test_that_extension_handles_alternate_CDCREC_race_encoding_with_custom_text() {
+
+        String patientWithDataForExtensions = "MSH|^~\\&|MIICEHRApplication|MIIC|MIIC|MIIC|201705130822||VXU^V04^VXU_V04|test1100|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS|^^^^^MIIC^SR^^^MIIC|MIIC\n"
+                // Special race with good code in known system but non-standard display text
+                + "PID|1||12345678^^^^MR|ALTID|Mouse^Mickey^J^III^^^|||M|Alias^Alias|2106-3^WHITE^CDCREC||USAA|||english|married||||DL00003333|||Orlando Disney Hospital|Y|2|USA||||\n";
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:149]: Similar to above, for this one I suggest:
`PID|1||12345678^^^^MR||Doe^Jane|||||2106-3^WHITE^CDCREC`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 4]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
+
+    }
+
+    @Test
+    void test_that_extension_handles_unknown_encoding() {
+
+        String patientWithDataForExtensions = "MSH|^~\\&|MIICEHRApplication|MIIC|MIIC|MIIC|201705130822||VXU^V04^VXU_V04|test1100|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS|^^^^^MIIC^SR^^^MIIC|MIIC\n"
+                // Special race with unknown system
+                + "PID|1||12345678^^^^MR|ALTID|Mouse^Mickey^J^III^^^|||M|Alias^Alias|White^WHITE^L||USAA|||english|married||||DL00003333|||Orlando Disney Hospital|Y|2|USA||||\n";
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:175]: Similar to above, for this one I suggest:
`PID|1||12345678^^^^MR||Doe^Jane|||||White^WHITE^L`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 5]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        assertThat(coding.getSystem()).containsIgnoringCase("urn:id:L");
+
+    }
+
+    @Test
+    void test_that_extension_with_valid_system_handles_missing_text_encoding() {
+
+        String patientWithDataForExtensions = "MSH|^~\\&|MIICEHRApplication|MIIC|MIIC|MIIC|201705130822||VXU^V04^VXU_V04|test1100|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS|^^^^^MIIC^SR^^^MIIC|MIIC\n"
+                // Special race with good code in known system but no display text
+                + "PID|1||12345678|ALTID|Mouse^Mickey^J^III^^^|||M|Alias^Alias|2106-3^^CDCREC||USAA|||english|married||||DL00003333|||Orlando Disney Hospital|Y|2|USA||||\n";
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:201]: Similar to above, for this one I suggest:
`PID|1||12345678^^^^MR||Doe^Jane|||||2106-3^^CDCREC`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 6]
```diff
@@ -140,6 +217,31 @@ void test_that_extension_handles_alternate_CDCREC_race_encoding() {
...
 
-    }    
+    }
+
+    @Test
+    void test_that_extension_with_invalid_code_and_valid_system_produces_failsafe_encoding() {
+
+        String patientWithDataForExtensions = "MSH|^~\\&|MIICEHRApplication|MIIC|MIIC|MIIC|201705130822||VXU^V04^VXU_V04|test1100|P|2.5.1|||AL|AL|||||Z22^CDCPHINVS|^^^^^MIIC^SR^^^MIIC|MIIC\n"
+                // Special race with bad code, but known system
+                + "PID|1||12345678|ALTID|Mouse^Mickey^J^III^^^|||M|Alias^Alias|2186-5^^CDCREC||USAA|||english|married||||DL00003333|||Orlando Disney Hospital|Y|2|USA||||\n";
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:227]: Similar to above, for this one I suggest:
`PID|1||12345678^^^^MR||Doe^Jane|||||2106-5^^CDCREC`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 7]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        CodeableConcept ccW = (CodeableConcept) extensions.get(0).getValue();
+        assertThat(ccW.hasCoding()).isTrue();
+        Coding coding = ccW.getCodingFirstRep();
+        assertThat(coding.hasDisplay()).isTrue();
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getDisplay()).hasToString("White");
+        assertThat(coding.getCode()).hasToString("2106-3");
+        assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
+
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:167]: Add checks:
```
assertThat(ccW.hasText().isTrue();
assertThat(ccW.getText().hasToString("WHITE");
```

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 8]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        assertThat(extensions.get(0).hasValue()).isTrue();
+        CodeableConcept ccW = (CodeableConcept) extensions.get(0).getValue();
+        assertThat(ccW.hasCoding()).isTrue();
+        Coding coding = ccW.getCodingFirstRep();
+        assertThat(coding.hasDisplay()).isTrue();
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getDisplay()).hasToString("WHITE");
+        assertThat(coding.getCode()).hasToString("White");
+        assertThat(coding.getSystem()).containsIgnoringCase("urn:id:L");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:192]: Add checks:
```
assertThat(ccW.hasText().isTrue();
assertThat(ccW.getText().hasToString("WHITE");
```

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 9]
```diff
@@ -140,6 +217,31 @@ void test_that_extension_handles_alternate_CDCREC_race_encoding() {
         assertThat(coding.getCode()).hasToString("2106-3");
         assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:219]: Add checks:
```
assertThat(ccW.hasText().isTrue();
assertThat(ccW.getText().hasToString("White");
```

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 10]
```diff
@@ -140,6 +217,31 @@ void test_that_extension_handles_alternate_CDCREC_race_encoding() {
...
+        assertThat(extensions.size()).isEqualTo(1);
+        assertThat(extensions.get(0).hasValue()).isTrue();
+        CodeableConcept ccW = (CodeableConcept) extensions.get(0).getValue();
+        assertThat(ccW.hasCoding()).isTrue();
+        Coding coding = ccW.getCodingFirstRep();
+        assertThat(coding.hasDisplay()).isFalse();
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getCode()).hasToString("2186-5");
+        assertThat(coding.getSystem()).containsIgnoringCase("urn:id:CDCREC");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:243]: Add check:
```
assertThat(ccW.hasText().isFalse();
```

Comment (LisaWellman) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:243]: coding.getSystem should be `http://terminology.hl7.org/CodeSystem/v3-Race` here

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 11]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
         assertThat(coding.hasCode()).isTrue();
         assertThat(coding.hasSystem()).isFalse();
         assertThat(coding.getCode()).hasToString("W");
-
     }
 
     @Test
     void test_that_extension_handles_alternate_CDCREC_race_encoding() {
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:119]: Java naming convention is camel case, and test methods are identified with @Test so do not have to have "test" in the name (though that can be useful).
For this one I suggest `testAlternateSystemCDCREC`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 12]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getDisplay()).hasToString("White");
+        assertThat(coding.getCode()).hasToString("2106-3");
+        assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
+
+    }
+
+    @Test
+    void test_that_extension_handles_alternate_CDCREC_race_encoding_with_custom_text() {
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:145]: Similar to above, for this one I suggest `testTextMismatchLookupSucceeds`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 13]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getDisplay()).hasToString("White");
+        assertThat(coding.getCode()).hasToString("2106-3");
+        assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
+
+    }
+
+    @Test
+    void test_that_extension_handles_unknown_encoding() {
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:171]: Similar to above, for this one I suggest `testUnknownSystem`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 14]
```diff
@@ -113,15 +113,92 @@ void test_that_extension_handles_text_only_race_correctly() {
...
+        assertThat(coding.hasCode()).isTrue();
+        assertThat(coding.hasSystem()).isTrue();
+        assertThat(coding.getDisplay()).hasToString("WHITE");
+        assertThat(coding.getCode()).hasToString("White");
+        assertThat(coding.getSystem()).containsIgnoringCase("urn:id:L");
+
+    }
+
+    @Test
+    void test_that_extension_with_valid_system_handles_missing_text_encoding() {
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:197]: Similar to above, for this one I suggest `testNoTextLookupSucceeds`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java [hunk 15]
```diff
@@ -140,6 +217,31 @@ void test_that_extension_handles_alternate_CDCREC_race_encoding() {
         assertThat(coding.getCode()).hasToString("2106-3");
         assertThat(coding.getSystem()).containsIgnoringCase("terminology.hl7.org/CodeSystem/v3-Race");
 
-    }    
+    }
+
+    @Test
+    void test_that_extension_with_invalid_code_and_valid_system_produces_failsafe_encoding() {
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:223]: Similar to above, for this one I suggest `testInvalidCodeNoText`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 12 above)
Comment (LisaWellman) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:145]: Although these tests are using the race extension, they are more about testing CodeableConcepts.  Let's move them into a new test class `CodeableConceptTest`

**Review by LisaWellman — COMMENTED** [review]
One main issue - in the last test the system is not correct

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 4 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:175]: Done.  Used `|Jane^TestPatientLastName|`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 5 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:201]: Done.  Used `|Jane^TestPatientLastName|`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 3 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:149]: Done.  Used `|Jane^TestPatientLastName|`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 2 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:141]: Done

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 1 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:123]: Done.  Used `|Jane^TestPatientLastName|`

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 11 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:119]: Made changes on all tests touched in this PR to have Java convention names.

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 12 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:145]: Moved several tests to CodeableConceptTest.java

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 14 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:197]: Renamed (slightly different).

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 15 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:223]: Renamed.

File: src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java (hunk 13 above)
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/hl7/segments/FHIRExtensionsTest.java:171]: Renamed

**Pull request merged**